### Import the required libraries

In [30]:
import pandas as pd
import os
import glob

### Task #1: Merge sales data for 12 months into one csv file

In [31]:
#We use glob to match the extension 'csv'
extension='csv'
all_sales_data_filenames= [i for i in glob.glob('Sales_Data/*{}'.format(extension))]
#print(all_sales_data_filenames)

In [32]:
#Combine all the files in a list 
all_sales_data=pd.concat([pd.read_csv(f) for f in all_sales_data_filenames])
#Export the combined data to csv
all_sales_data.to_csv("all_sales_data",index=False, encoding='utf-8-sig')

### Read the updated dataframe and have a general overview of the dataset

In [33]:
#Load the updated dataset
sales_data=pd.read_csv('all_sales_data')

In [34]:
#Print first 5 rows of the dataset
sales_data.head()

Order ID                Product Quantity Ordered Price Each      Order Date  \
0   248151  AA Batteries (4-pack)                4       3.84  09/17/19 14:44   
1   248152   USB-C Charging Cable                2      11.95  09/29/19 10:19   
2   248153   USB-C Charging Cable                1      11.95  09/16/19 17:48   
3   248154       27in FHD Monitor                1     149.99  09/27/19 07:52   
4   248155   USB-C Charging Cable                1      11.95  09/01/19 19:03   

                        Purchase Address  
0    380 North St, Los Angeles, CA 90001  
1           511 8th St, Austin, TX 73301  
2  151 Johnson St, Los Angeles, CA 90001  
3      355 Hickory St, Seattle, WA 98101  
4          125 5th St, Atlanta, GA 30301

In [35]:
#Describe the dataset
sales_data.describe()

Order ID               Product Quantity Ordered Price Each  \
count     186305                186305           186305     186305   
unique    178438                    20               10         24   
top     Order ID  USB-C Charging Cable                1      11.95   
freq         355                 21903           168552      21903   

        Order Date  Purchase Address  
count       186305            186305  
unique      142396            140788  
top     Order Date  Purchase Address  
freq           355               355

In [36]:
#Find more information about the dataset using the info() method
rows=sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB


In [37]:
#Print the number of rows and columns the dataset has

In [38]:
rows=len(sales_data)
columns=len(sales_data.columns)

print(f'The dataset has {rows} rows')
print(f'The dataset has {columns} columns')

The dataset has 186850 rows
The dataset has 6 columns


### Task #2: Cleaning the data 

#### Check for NaN/Missing values  

In [39]:
#Check if there ar any missing values in the dataframe
sales_data.isna().any().any()

True

We function has returned a boolean value of true, meaning there are columns with missing data in th dataset. Th next task is to see the count of the missing values

In [40]:
#Count the total number of missing values in the the dataset
sales_data.isna().sum().sum()

3270

From the result, we can see that our dataset has 3270 missing values

#### Drop rows with NaN values/ Missing values

In [41]:
sales_data=sales_data.dropna()
sales_data.isna().sum().sum()

0

#### Check for duplicate data in our dataset

In [42]:
#Find the number of duplicate rows in the dataset
sales_data.duplicated().sum()

618

#### Drop the duplicated rows

In [43]:
sales_data.drop_duplicates(inplace=True)

In [44]:
sales_data.duplicated().sum()

0

In [45]:
sales_data.head()

Order ID                Product Quantity Ordered Price Each      Order Date  \
0   248151  AA Batteries (4-pack)                4       3.84  09/17/19 14:44   
1   248152   USB-C Charging Cable                2      11.95  09/29/19 10:19   
2   248153   USB-C Charging Cable                1      11.95  09/16/19 17:48   
3   248154       27in FHD Monitor                1     149.99  09/27/19 07:52   
4   248155   USB-C Charging Cable                1      11.95  09/01/19 19:03   

                        Purchase Address  
0    380 North St, Los Angeles, CA 90001  
1           511 8th St, Austin, TX 73301  
2  151 Johnson St, Los Angeles, CA 90001  
3      355 Hickory St, Seattle, WA 98101  
4          125 5th St, Atlanta, GA 30301

#### Check the data types and make sure they conform with their respective columns

In [46]:
#Check the data type of columns in the dataset
sales_data.dtypes

Order ID            object
Product             object
Quantity Ordered    object
Price Each          object
Order Date          object
Purchase Address    object
dtype: object

From the above results, we can see that all the columns are of type 'object'. However, some of the columns need to be cleaned to specific data types. They include price, and date.

#### Resolve the data type issue

In [51]:
#Convert Order ID and Price Each columns to numeric data type
sales_data[['Order ID','Price Each']]=sales_data[['Order ID',
                                    'Price Each']].apply(pd.to_numeric, errors='coerce')

In [52]:
sales_data.dtypes

Order ID            float64
Product              object
Quantity Ordered     object
Price Each          float64
Order Date           object
Purchase Address     object
dtype: object